libs/langchain-openai/src/chat_models.ts

```diff
@@ -843,4 +854,118 @@ export class ChatOpenAI<
+   * @param {boolean | undefined} includeRaw Whether to include the raw output in the result. Defaults to false.
+   * @returns {Runnable<RunInput, RunOutput, CallOptions>} A new runnable that calls the LLM with structured output.
+   */
+  withStructuredOutput<
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    RunInput = any,
+    // prettier-ignore
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    RunOutput extends z.ZodObject<any, any, any, any> = z.ZodObject<any, any, any, any>
+  >({
+    schema,
+    name,
+    method = "functionCalling",
+    includeRaw = false,
+  }: {
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    schema: z.ZodEffects<RunOutput> | Record<string, any>;
+    name: string;
+    method?: "functionCalling" | "jsonMode";
+    includeRaw?: boolean;
+  }): Runnable<RunInput, RunOutput> {
+    let llm: Runnable;
+    let outputParser: JsonOutputKeyToolsParser | JsonOutputParser<RunOutput>;
+
+    if (method === "jsonMode") {
+      llm = this.bind({
+        response_format: { type: "json_object" },
+      } as Partial<CallOptions>);
+      outputParser = new JsonOutputParser<RunOutput>();
+    } else {
+      // Is function calling
+      if (isZodSchema(schema)) {
+        class TmpClass extends StructuredTool {
+          schema = schema as z.ZodEffects<RunOutput>;
+
+          description = schema.description;
+
+          // We need this because TypeScript can not infer that name will not be undefined
+          // even though there is a check above.
+          name = name ?? "";
+
+          async _call(input: RunInput) {
+            return JSON.stringify(input);
+          }
+        }
+        llm = this.bind({
+          tools: [new TmpClass()],
+          tool_choice: "auto",
+        } as unknown as Partial<CallOptions>);
+        outputParser = new JsonOutputKeyToolsParser({
+          returnSingle: true,
+          keyName: name,
+        });
+      } else {
+        llm = this.bind({
+          tools: [
+            {
+              type: "function" as const,
+              function: {
+                name,
+                parameters: schema,
+              },
+            },
+          ],
+          tool_choice: "auto",
+        } as unknown as Partial<CallOptions>);
+        outputParser = new JsonOutputKeyToolsParser({
+          returnSingle: true,
+          keyName: name,
+        });
+      }
+    }
+
+    if (!includeRaw) {
+      return llm.pipe(outputParser);
+    }
+
+    const parserAssign = RunnablePassthrough.assign({
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      parsed: (input: any) => outputParser.invoke(input.raw),
+    });
+    const parserNone = RunnablePassthrough.assign({
+      parsed: () => null,
+    });
+    const parsedWithFallback = parserAssign.withFallbacks({
+      fallbacks: [parserNone],
+    });
+    return new RunnableMap({
```

Comment:
Just use `RunnableSequence.from()` - not really intended to use the map constructor directly as it's a bit cumbersome